src/itwinai/torch/distributed.py

```diff
@@ -1015,3 +1039,93 @@ def gather(self, tensor: torch.Tensor, dst_rank: int = 0):
             list[Any]: input object wrapped in a list.
         """
         return [tensor]
+
+
+class RayDDPStrategy(TorchDDPStrategy):
+    """A distributed data-parallel (DDP) strategy using Ray Train for PyTorch training."""
+
+    def __init__(self) -> None:
+        _initialize_ray()
+
+    def init(self) -> None:
+        self.is_initialized = True
+
+    @check_initialized
+    def global_world_size(self) -> int:
+        return ray.train.get_context().get_world_size()
+
+    @check_initialized
+    def local_world_size(self) -> int:
+        return ray.train.get_context().get_local_world_size()
+
+    @check_initialized
+    def global_rank(self) -> int:
+        return ray.train.get_context().get_world_rank()
+
+    @check_initialized
+    def local_rank(self) -> int:
+        return ray.train.get_context().get_local_rank()
+
+    @check_initialized
+    def distributed(
+        self,
+        model: nn.Module,
+        optimizer: Optimizer,
+        lr_scheduler: Optional[LRScheduler] = None,
+    ) -> Tuple[nn.Module, Optimizer, LRScheduler | None]:
+        model = ray.train.torch.prepare_model(model)
+
+        return model, optimizer, lr_scheduler
+
+    @check_initialized
+    def create_dataloader(
+        self,
+        dataset: Dataset[T_co],
+        batch_size: Optional[int] = 1,
+        shuffle: Optional[bool] = None,
+        sampler: Union[Sampler, Iterable, None] = None,
+        batch_sampler: Union[Sampler[List], Iterable[List], None] = None,
+        num_workers: int = 0,
+        collate_fn: Optional[_collate_fn_t] = None,
+        pin_memory: bool = False,
+        drop_last: bool = False,
+        timeout: float = 0,
+        worker_init_fn: Optional[_worker_init_fn_t] = None,
+        multiprocessing_context=None,
+        generator=None,
+        *,
+        prefetch_factor: Optional[int] = None,
+        persistent_workers: bool = False,
+        pin_memory_device: str = "",
+    ):
+        dataloader = super().create_dataloader(
```

Comment:
be careful that the `super().create_dataloader(...)` method will create a data loader *already* with a distributed sampler, which may not be what you want to do when using Ray (e.g., because it may not be needed). If you don't need this, replace `super().create_dataloader(...)` simply with `Dataloader`

Reply:
Good catch, thanks!